diff --git a/src/Lazy.ts b/src/Lazy.ts
index eab4a61..f3f6ace 100644
--- a/src/Lazy.ts
+++ b/src/Lazy.ts
@@ -3,7 +3,7 @@ import type {
   AnyObject,
   ISchema,
   ValidateOptions,
-  NestedTestConfig,
+  InternalOptions,
   InferType,
 } from './types';
 import type { ResolveOptions } from './Condition';
@@ -117,15 +117,21 @@ class Lazy<T, TContext = AnyObject, TFlags extends Flags = any>
     return this._resolve(value, options).cast(value, options as any);
   }
 
-  asNestedTest(config: NestedTestConfig) {
-    let { key, index, parent, options } = config;
-    let value = parent[index ?? key!];
+  asNestedTest(
+    key?: string,
+    index?: number,
+    parent?: any,
+    parentPath?: string,
+    originalParent?: any,
+    options: InternalOptions<any> = {},
+  ) {
+    let value = parent![index ?? key!];
 
     return this._resolve(value, {
       ...options,
       value,
       parent,
-    }).asNestedTest(config);
+    }).asNestedTest(key, index, parent, parentPath, originalParent, options);
   }
 
   validate(value: any, options?: ValidateOptions<TContext>): Promise<T> {
diff --git a/src/Reference.ts b/src/Reference.ts
index c04ff16..96889fc 100644
--- a/src/Reference.ts
+++ b/src/Reference.ts
@@ -62,15 +62,8 @@ export default class Reference<TValue = unknown> {
     return result;
   }
 
-  /**
-   *
-   * @param {*} value
-   * @param {Object} options
-   * @param {Object=} options.context
-   * @param {Object=} options.parent
-   */
-  cast(value: any, options?: { parent?: {}; context?: {} }) {
-    return this.getValue(value, options?.parent, options?.context);
+  cast(value: any, parent?: any, context?: any) {
+    return this.getValue(value, parent, context);
   }
 
   resolve() {
diff --git a/src/array.ts b/src/array.ts
index 29f030b..657e2b9 100644
--- a/src/array.ts
+++ b/src/array.ts
@@ -126,22 +126,22 @@ export default class ArraySchema<
       // #950 Ensure that sparse array empty slots are validated
       let tests: RunTest[] = new Array(value.length);
       for (let index = 0; index < value.length; index++) {
-        tests[index] = innerType!.asNestedTest({
-          options,
+        tests[index] = innerType!.asNestedTest(
+          undefined,
           index,
-          parent: value,
-          parentPath: options.path,
-          originalParent: options.originalValue ?? _value,
-        });
+          value,
+          options.path,
+          options.originalValue ?? _value,
+          options,
+        );
       }
 
       this.runTests(
-        {
-          value,
-          tests,
-          originalValue: options.originalValue ?? _value,
-          options,
-        },
+        tests,
+        value,
+        options.originalValue ?? _value,
+        undefined,
+        options,
         panic,
         (innerTypeErrors) => next(innerTypeErrors.concat(arrayErrors), value),
       );
diff --git a/src/object.ts b/src/object.ts
index 94dc817..b6fcfe3 100644
--- a/src/object.ts
+++ b/src/object.ts
@@ -205,7 +205,13 @@ export default class ObjectSchema<
 
         fieldValue =
           !options.__validating || !strict
-            ? (field as ISchema<any>).cast(inputValue, innerOptions)
+            ? Reference.isRef(field)
+              ? field.cast(
+                  inputValue,
+                  innerOptions.parent,
+                  innerOptions.context,
+                )
+              : (field as ISchema<any>).cast(inputValue, innerOptions)
             : inputValue;
 
         if (fieldValue !== undefined) {
@@ -261,18 +267,23 @@ export default class ObjectSchema<
         }
 
         tests.push(
-          field.asNestedTest({
-            options,
+          field.asNestedTest(
             key,
-            parent: value,
-            parentPath: options.path,
-            originalParent: originalValue,
-          }),
+            undefined,
+            value,
+            options.path,
+            originalValue,
+            options,
+          ),
         );
       }
 
       this.runTests(
-        { tests, value, originalValue, options },
+        tests,
+        value,
+        originalValue,
+        undefined,
+        options,
         panic,
         (fieldErrors) => {
           next(fieldErrors.sort(this._sortErrors).concat(objectErrors), value);
@@ -325,9 +336,9 @@ export default class ObjectSchema<
     );
   }
 
-  protected _getDefault(options?: ResolveOptions<TContext>) {
+  protected _getDefault(value?: any, parent?: any, context?: any) {
     if ('default' in this.spec) {
-      return super._getDefault(options);
+      return super._getDefault(value, parent, context);
     }
 
     // if there is no default set invent one
@@ -339,18 +350,20 @@ export default class ObjectSchema<
     this._nodes.forEach((key) => {
       const field = this.fields[key] as any;
 
-      let innerOptions = options;
-      if (innerOptions?.value) {
-        innerOptions = {
-          ...innerOptions,
-          parent: innerOptions.value,
-          value: innerOptions.value[key],
-        };
+      let innerValue: any;
+      let innerParent: any;
+      if (value) {
+        innerValue = value[key];
+        innerParent = value;
       }
 
       dft[key] =
         field && 'getDefault' in field
-          ? field.getDefault(innerOptions)
+          ? field.getDefault({
+              value: innerValue,
+              parent: innerParent,
+              context,
+            })
           : undefined;
     });
 
diff --git a/src/schema.ts b/src/schema.ts
index 2d20330..7e10627 100644
--- a/src/schema.ts
+++ b/src/schema.ts
@@ -10,7 +10,6 @@ import createValidation, {
   TestConfig,
   NextCallback,
   PanicCallback,
-  TestOptions,
   resolveParams,
 } from './util/createValidation';
 import printValue from './util/printValue';
@@ -23,7 +22,6 @@ import {
   InternalOptions,
   ExtraParams,
   ISchema,
-  NestedTestConfig,
   DefaultThunk,
 } from './types';
 
@@ -102,7 +100,11 @@ export interface CastOptionalityOptions<C = {}>
 }
 
 export type RunTest = (
-  opts: TestOptions,
+  value: any,
+  path: string | undefined,
+  options: InternalOptions,
+  originalValue: any,
+  schema: AnySchema,
   panic: PanicCallback,
   next: NextCallback,
 ) => void;
@@ -445,13 +447,11 @@ export default abstract class Schema<
     }
 
     this.runTests(
-      {
-        path,
-        value,
-        originalValue,
-        options,
-        tests: initialTests,
-      },
+      initialTests,
+      value,
+      originalValue,
+      path,
+      options,
       panic,
       (initialErrors) => {
         // even if we aren't ending early we can't proceed further if the types aren't correct
@@ -460,13 +460,11 @@ export default abstract class Schema<
         }
 
         this.runTests(
-          {
-            path,
-            value,
-            originalValue,
-            options,
-            tests: this.tests,
-          },
+          this.tests,
+          value,
+          originalValue,
+          path,
+          options,
           panic,
           next,
         );
@@ -479,12 +477,15 @@ export default abstract class Schema<
    * schema validate result.
    */
   protected runTests(
-    runOptions: TestRunOptions,
+    tests: RunTest[],
+    value: any,
+    originalValue: any,
+    path: string | undefined,
+    options: InternalOptions,
     panic: (err: Error, value: unknown) => void,
     next: (errors: ValidationError[], value: unknown) => void,
   ): void {
     let fired = false;
-    let { tests, value, originalValue, path, options } = runOptions;
 
     let panicOnce = (arg: Error) => {
       if (fired) return;
@@ -503,18 +504,12 @@ export default abstract class Schema<
 
     if (!count) return nextOnce([]);
 
-    let args = {
-      value,
-      originalValue,
-      path,
-      options,
-      schema: this,
-    };
-
     for (let i = 0; i < tests.length; i++) {
       const test = tests[i];
 
-      test(args!, panicOnce, function finishTestRun(err) {
+      test(value, path, options, originalValue, this, panicOnce, function finishTestRun(
+        err,
+      ) {
         if (err) {
           Array.isArray(err)
             ? nestedErrors.push(...err)
@@ -527,21 +522,21 @@ export default abstract class Schema<
     }
   }
 
-  asNestedTest({
-    key,
-    index,
-    parent,
-    parentPath,
-    originalParent,
-    options,
-  }: NestedTestConfig): RunTest {
+  asNestedTest(
+    key?: string,
+    index?: number,
+    parent?: any,
+    parentPath?: string,
+    originalParent?: any,
+    options: InternalOptions<any> = {},
+  ): RunTest {
     const k = key ?? index;
     if (k == null) {
       throw TypeError('Must include `key` or `index` for nested validations');
     }
 
     const isIndex = typeof k === 'number';
-    let value = parent[k];
+    let value = parent![k];
 
     const testOptions = {
       ...options,
@@ -551,7 +546,7 @@ export default abstract class Schema<
       strict: true,
       parent,
       value,
-      originalValue: originalParent[k],
+      originalValue: originalParent![k],
       // FIXME: tests depend on `index` being passed around deeply,
       //   we should not let the options.key/index bleed through
       key: undefined,
@@ -563,8 +558,15 @@ export default abstract class Schema<
           : (parentPath ? `${parentPath}.` : '') + key,
     };
 
-    return (_: any, panic, next) =>
-      this.resolve(testOptions)._validate(value, testOptions, panic, next);
+    return (
+      _value: any,
+      _path: string | undefined,
+      _options: InternalOptions,
+      _originalValue: any,
+      _schema: AnySchema,
+      panic: PanicCallback,
+      next: NextCallback,
+    ) => this.resolve(testOptions)._validate(value, testOptions, panic, next);
   }
 
   validate(
@@ -655,7 +657,7 @@ export default abstract class Schema<
     }
   }
 
-  protected _getDefault(options?: ResolveOptions<TContext>) {
+  protected _getDefault(value?: any, parent?: any, context?: any) {
     let defaultValue = this.spec.default;
 
     if (defaultValue == null) {
@@ -663,7 +665,7 @@ export default abstract class Schema<
     }
 
     return typeof defaultValue === 'function'
-      ? defaultValue.call(this, options)
+      ? defaultValue.call(this, { value, parent, context })
       : cloneDeep(defaultValue);
   }
 
@@ -672,7 +674,7 @@ export default abstract class Schema<
     // If schema is defaulted we know it's at least not undefined
   ): TDefault {
     let schema = this.resolve(options || {});
-    return schema._getDefault(options);
+    return schema._getDefault(options?.value, options?.parent, options?.context);
   }
 
   default(def: DefaultThunk<any>): any {
@@ -963,7 +965,12 @@ export default abstract class Schema<
         .map((fn) => {
           const params =
             fn.OPTIONS!.params && options
-              ? resolveParams({ ...fn.OPTIONS!.params }, options)
+              ? resolveParams(
+                  { ...fn.OPTIONS!.params },
+                  options.value,
+                  options.parent,
+                  options.context,
+                )
               : fn.OPTIONS!.params;
 
           return {
diff --git a/src/tuple.ts b/src/tuple.ts
index 54bb8a9..9bbf707 100644
--- a/src/tuple.ts
+++ b/src/tuple.ts
@@ -143,22 +143,22 @@ export default class TupleSchema<
 
       let tests: RunTest[] = [];
       for (let [index, itemSchema] of itemTypes.entries()) {
-        tests[index] = itemSchema!.asNestedTest({
-          options,
+        tests[index] = itemSchema!.asNestedTest(
+          undefined,
           index,
-          parent: value,
-          parentPath: options.path,
-          originalParent: options.originalValue ?? _value,
-        });
+          value,
+          options.path,
+          options.originalValue ?? _value,
+          options,
+        );
       }
 
       this.runTests(
-        {
-          value,
-          tests,
-          originalValue: options.originalValue ?? _value,
-          options,
-        },
+        tests,
+        value,
+        options.originalValue ?? _value,
+        undefined,
+        options,
         panic,
         (innerTypeErrors) => next(innerTypeErrors.concat(tupleErrors), value),
       );
diff --git a/src/types.ts b/src/types.ts
index 9d64abf..2b5b504 100644
--- a/src/types.ts
+++ b/src/types.ts
@@ -23,7 +23,14 @@ export interface ISchema<T, C = any, F extends Flags = any, D = any> {
 
   validate(value: any, options?: ValidateOptions<C>): Promise<T>;
 
-  asNestedTest(config: NestedTestConfig): Test;
+  asNestedTest(
+    key?: string,
+    index?: number,
+    parent?: any,
+    parentPath?: string,
+    originalParent?: any,
+    options?: InternalOptions<any>,
+  ): Test;
 
   describe(options?: ResolveOptions<C>): SchemaFieldDescription;
   resolve(options: ResolveOptions<C>): ISchema<T, C, F>;
diff --git a/src/util/createValidation.ts b/src/util/createValidation.ts
index 9243a44..a97d0e6 100644
--- a/src/util/createValidation.ts
+++ b/src/util/createValidation.ts
@@ -10,7 +10,6 @@ import {
 import Reference from '../Reference';
 import type { AnySchema } from '../schema';
 import isAbsent from './isAbsent';
-import { ResolveOptions } from '../Condition';
 
 export type PanicCallback = (err: Error) => void;
 
@@ -61,7 +60,11 @@ export type TestConfig<TValue = unknown, TContext = {}> = {
 };
 
 export type Test = ((
-  opts: TestOptions,
+  value: any,
+  path: string | undefined,
+  options: InternalOptions,
+  originalValue: any,
+  schema: AnySchema,
   panic: PanicCallback,
   next: NextCallback,
 ) => void) & {
@@ -76,7 +79,11 @@ export default function createValidation(config: {
   skipAbsent?: boolean;
 }) {
   function validate<TSchema extends AnySchema = AnySchema>(
-    { value, path = '', options, originalValue, schema }: TestOptions<TSchema>,
+    value: any,
+    path: string | undefined,
+    options: InternalOptions,
+    originalValue: any,
+    schema: TSchema,
     panic: PanicCallback,
     next: NextCallback,
   ) {
@@ -87,7 +94,9 @@ export default function createValidation(config: {
       abortEarly = schema.spec.abortEarly,
       disableStackTrace = schema.spec.disableStackTrace,
     } = options;
-    const resolveOptions = { value, parent, context };
+
+    path = path || '';
+
     function createError(overrides: CreateErrorOptions = {}) {
       const nextParams = resolveParams(
         {
@@ -100,7 +109,9 @@ export default function createValidation(config: {
           ...params,
           ...overrides.params,
         },
-        resolveOptions,
+        value,
+        parent,
+        context,
       );
 
       const error = new ValidationError(
@@ -123,7 +134,7 @@ export default function createValidation(config: {
       from: options.from,
       createError,
       resolve<T>(item: T | Reference<T>) {
-        return resolveMaybeRef(item, resolveOptions);
+        return resolveMaybeRef(item, value, parent, context);
       },
       options,
       originalValue,
@@ -175,20 +186,25 @@ export default function createValidation(config: {
 // Warning: mutates the input
 export function resolveParams<T extends ExtraParams>(
   params: T,
-  options: ResolveOptions,
+  value?: any,
+  parent?: any,
+  context?: any,
 ) {
   if (!params) return params;
 
   type Keys = (keyof typeof params)[];
   for (const key of Object.keys(params) as Keys) {
-    params[key] = resolveMaybeRef(params[key], options);
+    params[key] = resolveMaybeRef(params[key], value, parent, context);
   }
 
   return params;
 }
 
-function resolveMaybeRef<T>(item: T | Reference<T>, options: ResolveOptions) {
-  return Ref.isRef(item)
-    ? item.getValue(options.value, options.parent, options.context)
-    : item;
+function resolveMaybeRef<T>(
+  item: T | Reference<T>,
+  value?: any,
+  parent?: any,
+  context?: any,
+) {
+  return Ref.isRef(item) ? item.getValue(value, parent, context) : item;
 }
