diff --git a/numpy/_core/src/highway b/numpy/_core/src/highway
index 37c08e5528..ee36c83712 160000
--- a/numpy/_core/src/highway
+++ b/numpy/_core/src/highway
@@ -1 +1 @@
-Subproject commit 37c08e5528f63ead9c7e4fd99ba454c1b1a3e3f7
+Subproject commit ee36c837129310be19c17c9108c6dc3f6ae06942
diff --git a/numpy/ma/core.py b/numpy/ma/core.py
index 8a85960f62..7e320e2119 100644
--- a/numpy/ma/core.py
+++ b/numpy/ma/core.py
@@ -950,6 +950,15 @@ def __init__(self, ufunc):
     def __str__(self):
         return f"Masked version of {self.f}"
 
+    def reduce(self, target, axis=0, dtype=None):
+        raise NotImplementedError("reduce is not supported")
+
+    def outer(self, a, b):
+        raise NotImplementedError("outer is not supported")
+
+    def accumulate(self, target, axis=0):
+        raise NotImplementedError("accumulate is not supported")
+
 
 class _MaskedUnaryOperation(_MaskedUFunc):
     """
diff --git a/vendored-meson/meson b/vendored-meson/meson
index e72c717199..5d5a3d478d 160000
--- a/vendored-meson/meson
+++ b/vendored-meson/meson
@@ -1 +1 @@
-Subproject commit e72c717199fa18d34020c7c97f9de3f388c5e055
+Subproject commit 5d5a3d478da115c812be77afa651db2492d52171
