diff --git a/jarviscli/plugins/reminder.py b/jarviscli/plugins/reminder.py
index 8de5b82..fb0f03d 100644
--- a/jarviscli/plugins/reminder.py
+++ b/jarviscli/plugins/reminder.py
@@ -124,7 +124,7 @@ class TagBase:
                 self.clean_up_entry(jarvis, entry)
 
             tags_store.items = []
-            self.save_data(jarvis, tags_store.items)
+            tags_store.save()
 
             jarvis.say("ok")
             return
