diff --git a/spectrwm.c b/spectrwm.c
index f76d55d..8fc35e7 100644
--- a/spectrwm.c
+++ b/spectrwm.c
@@ -9331,65 +9331,13 @@ update_region_mapping(struct swm_region *r)
 		if (!HIDDEN(w) && (!mof || (w && wf && w->main == wf->main) ||
 		    WINDOCK(w))) {
 			/* Map the window in one step and restamp WM_STATE as NORMAL. */
-			DNPRINTF(SWM_D_EVENT, "win %#x, mapped: %s\n",
-			    w->id, YESNO(w->mapped));
-			if (w->frame == XCB_WINDOW_NONE) {
-				DNPRINTF(SWM_D_EVENT,
-				    "skip win %#x; not reparented\n", w->id);
-			} else if (!w->mapped) {
-				xcb_map_window(conn, w->frame);
-				xcb_map_window(conn, w->id);
-				if (w->debug != XCB_WINDOW_NONE)
-					xcb_map_window(conn, w->debug);
-				w->mapping += 2;
-				w->mapped = true;
-				DNPRINTF(SWM_D_EVENT,
-				    "win %#x, state: %s(%u)\n", WINID(w),
-				    get_wm_state_label(XCB_ICCCM_WM_STATE_NORMAL),
-				    XCB_ICCCM_WM_STATE_NORMAL);
-				if (w->id != XCB_WINDOW_NONE) {
-					uint32_t	data[2] = {
-					    XCB_ICCCM_WM_STATE_NORMAL,
-					    XCB_WINDOW_NONE
-					};
-					w->state = XCB_ICCCM_WM_STATE_NORMAL;
-					xcb_change_property(conn,
-					    XCB_PROP_MODE_REPLACE, w->id,
-					    a_state, a_state, 32, 2, data);
-				}
-			}
+			map_window(w);
 		}
 	TAILQ_FOREACH(w, &ws->winlist, entry)
 		if (HIDDEN(w) || (mof && !(w && wf && w->main == wf->main) &&
 		    !WINDOCK(w))) {
 			/* Unmap the window in one step and restamp WM_STATE as ICONIC. */
-			DNPRINTF(SWM_D_EVENT, "win %#x, mapped: %s\n", w->id,
-			    YESNO(w->mapped));
-			if (w->frame == XCB_WINDOW_NONE) {
-				DNPRINTF(SWM_D_EVENT,
-				    "skip win %#x; not reparented\n", w->id);
-			} else if (w->mapped) {
-				if (w->debug != XCB_WINDOW_NONE)
-					xcb_unmap_window(conn, w->debug);
-				xcb_unmap_window(conn, w->id);
-				xcb_unmap_window(conn, w->frame);
-				w->unmapping += 2;
-				w->mapped = false;
-				DNPRINTF(SWM_D_EVENT,
-				    "win %#x, state: %s(%u)\n", WINID(w),
-				    get_wm_state_label(XCB_ICCCM_WM_STATE_ICONIC),
-				    XCB_ICCCM_WM_STATE_ICONIC);
-				if (w->id != XCB_WINDOW_NONE) {
-					uint32_t	data[2] = {
-					    XCB_ICCCM_WM_STATE_ICONIC,
-					    XCB_WINDOW_NONE
-					};
-					w->state = XCB_ICCCM_WM_STATE_ICONIC;
-					xcb_change_property(conn,
-					    XCB_PROP_MODE_REPLACE, w->id,
-					    a_state, a_state, 32, 2, data);
-				}
-			}
+			unmap_window(w);
 		}
 }
 
