diff --git a/src/jinja2/compiler.py b/src/jinja2/compiler.py
index a4ff6a1b..de9941d0 100644
--- a/src/jinja2/compiler.py
+++ b/src/jinja2/compiler.py
@@ -14,6 +14,7 @@ from . import nodes
 from .exceptions import TemplateAssertionError
 from .idtracking import Symbols
 from .idtracking import VAR_LOAD_ALIAS
+from .idtracking import VAR_LOAD_OVERLAY
 from .idtracking import VAR_LOAD_PARAMETER
 from .idtracking import VAR_LOAD_RESOLVE
 from .idtracking import VAR_LOAD_UNDEFINED
@@ -592,6 +593,13 @@ class CodeGenerator(NodeVisitor):
                 self.writeline(f"{target} = {param}")
             elif action == VAR_LOAD_UNDEFINED:
                 undefs.append(target)
+            elif action == VAR_LOAD_OVERLAY:
+                self.writeline(f"{target} = {self.get_resolve_func()}({param!r})")
+                if frame.toplevel:
+                    self.writeline(f"context.exported_vars.add({param!r})")
+                    ref = frame.symbols.find_ref(param)
+                    if ref is not None:
+                        self.writeline(f"context.vars[{param!r}] = {ref}")
             else:
                 raise NotImplementedError("unknown load instruction")
         if undefs:
diff --git a/src/jinja2/idtracking.py b/src/jinja2/idtracking.py
index e6dd8cd1..04e766da 100644
--- a/src/jinja2/idtracking.py
+++ b/src/jinja2/idtracking.py
@@ -10,6 +10,7 @@ VAR_LOAD_PARAMETER = "param"
 VAR_LOAD_RESOLVE = "resolve"
 VAR_LOAD_ALIAS = "alias"
 VAR_LOAD_UNDEFINED = "undefined"
+VAR_LOAD_OVERLAY = "overlay"
 
 
 def find_symbols(
