diff --git a/pandas/core/reshape/merge.py b/pandas/core/reshape/merge.py
index a1e003c200..7355dc1484 100644
--- a/pandas/core/reshape/merge.py
+++ b/pandas/core/reshape/merge.py
@@ -149,6 +149,14 @@ def merge(
     indicator: str | bool = False,
     validate: str | None = None,
 ) -> DataFrame:
+    if not is_list_like(suffixes, allow_sets=False) or isinstance(suffixes, dict):
+        raise TypeError(
+            f"Passing 'suffixes' as a {type(suffixes)}, is not supported. "
+            "Provide 'suffixes' as a tuple instead."
+        )
+    left_suffix, right_suffix = suffixes
+    suffix_check_dups = True
+
     left_df = _validate_operand(left)
     right_df = _validate_operand(right)
     if how == "cross":
@@ -161,7 +169,9 @@ def merge(
             left_index=left_index,
             right_index=right_index,
             sort=sort,
-            suffixes=suffixes,
+            left_suffix=left_suffix,
+            right_suffix=right_suffix,
+            suffix_check_dups=suffix_check_dups,
             indicator=indicator,
             validate=validate,
             copy=copy,
@@ -177,7 +187,9 @@ def merge(
             left_index=left_index,
             right_index=right_index,
             sort=sort,
-            suffixes=suffixes,
+            left_suffix=left_suffix,
+            right_suffix=right_suffix,
+            suffix_check_dups=suffix_check_dups,
             indicator=indicator,
             validate=validate,
         )
@@ -193,7 +205,9 @@ def _cross_merge(
     left_index: bool = False,
     right_index: bool = False,
     sort: bool = False,
-    suffixes: Suffixes = ("_x", "_y"),
+    left_suffix: str | None = "_x",
+    right_suffix: str | None = "_y",
+    suffix_check_dups: bool = True,
     copy: bool | None = None,
     indicator: str | bool = False,
     validate: str | None = None,
@@ -230,7 +244,7 @@ def _cross_merge(
         left_index=left_index,
         right_index=right_index,
         sort=sort,
-        suffixes=suffixes,
+        suffixes=(left_suffix, right_suffix),
         indicator=indicator,
         validate=validate,
         copy=copy,
@@ -400,6 +414,9 @@ def merge_ordered(
     9   e       3     b     3.0
     """
 
+    left_suffix, right_suffix = suffixes
+    suffix_check_dups = True
+
     def _merger(x, y) -> DataFrame:
         # perform the ordered merge operation
         op = _OrderedMerge(
@@ -408,7 +425,9 @@ def merge_ordered(
             on=on,
             left_on=left_on,
             right_on=right_on,
-            suffixes=suffixes,
+            left_suffix=left_suffix,
+            right_suffix=right_suffix,
+            suffix_check_dups=suffix_check_dups,
             fill_method=fill_method,
             how=how,
         )
@@ -688,6 +707,9 @@ def merge_asof(
     3 2016-05-25 13:30:00.048   GOOG  720.92       100     NaN     NaN
     4 2016-05-25 13:30:00.048   AAPL   98.00       100     NaN     NaN
     """
+    left_suffix, right_suffix = suffixes
+    suffix_check_dups = True
+
     op = _AsOfMerge(
         left,
         right,
@@ -699,7 +721,9 @@ def merge_asof(
         by=by,
         left_by=left_by,
         right_by=right_by,
-        suffixes=suffixes,
+        left_suffix=left_suffix,
+        right_suffix=right_suffix,
+        suffix_check_dups=suffix_check_dups,
         how="asof",
         tolerance=tolerance,
         allow_exact_matches=allow_exact_matches,
@@ -726,7 +750,9 @@ class _MergeOperation:
     left_index: bool
     right_index: bool
     sort: bool
-    suffixes: Suffixes
+    left_suffix: str | None
+    right_suffix: str | None
+    suffix_check_dups: bool
     copy: bool
     indicator: str | bool
     validate: str | None
@@ -745,7 +771,9 @@ class _MergeOperation:
         left_index: bool = False,
         right_index: bool = False,
         sort: bool = True,
-        suffixes: Suffixes = ("_x", "_y"),
+        left_suffix: str | None = "_x",
+        right_suffix: str | None = "_y",
+        suffix_check_dups: bool = True,
         indicator: str | bool = False,
         validate: str | None = None,
     ) -> None:
@@ -757,7 +785,9 @@ class _MergeOperation:
 
         self.on = com.maybe_make_list(on)
 
-        self.suffixes = suffixes
+        self.left_suffix = left_suffix
+        self.right_suffix = right_suffix
+        self.suffix_check_dups = suffix_check_dups
         self.sort = sort or how == "outer"
 
         self.left_index = left_index
@@ -838,7 +868,11 @@ class _MergeOperation:
         right = self.right[:]
 
         llabels, rlabels = _items_overlap_with_suffix(
-            self.left._info_axis, self.right._info_axis, self.suffixes
+            self.left._info_axis,
+            self.right._info_axis,
+            self.left_suffix,
+            self.right_suffix,
+            self.suffix_check_dups,
         )
 
         if left_indexer is not None and not is_range_indexer(left_indexer, len(left)):
@@ -1904,7 +1938,9 @@ class _OrderedMerge(_MergeOperation):
         right_on: IndexLabel | None = None,
         left_index: bool = False,
         right_index: bool = False,
-        suffixes: Suffixes = ("_x", "_y"),
+        left_suffix: str | None = "_x",
+        right_suffix: str | None = "_y",
+        suffix_check_dups: bool = True,
         fill_method: str | None = None,
         how: JoinHow | Literal["asof"] = "outer",
     ) -> None:
@@ -1919,7 +1955,9 @@ class _OrderedMerge(_MergeOperation):
             right_index=right_index,
             right_on=right_on,
             how=how,
-            suffixes=suffixes,
+            left_suffix=left_suffix,
+            right_suffix=right_suffix,
+            suffix_check_dups=suffix_check_dups,
             sort=True,  # factorize sorts
         )
 
@@ -1972,7 +2010,9 @@ class _AsOfMerge(_OrderedMerge):
         by=None,
         left_by=None,
         right_by=None,
-        suffixes: Suffixes = ("_x", "_y"),
+        left_suffix: str | None = "_x",
+        right_suffix: str | None = "_y",
+        suffix_check_dups: bool = True,
         how: Literal["asof"] = "asof",
         tolerance=None,
         allow_exact_matches: bool = True,
@@ -2007,7 +2047,9 @@ class _AsOfMerge(_OrderedMerge):
             left_index=left_index,
             right_index=right_index,
             how=how,
-            suffixes=suffixes,
+            left_suffix=left_suffix,
+            right_suffix=right_suffix,
+            suffix_check_dups=suffix_check_dups,
             fill_method=None,
         )
 
@@ -2695,7 +2737,11 @@ def _validate_operand(obj: DataFrame | Series) -> DataFrame:
 
 
 def _items_overlap_with_suffix(
-    left: Index, right: Index, suffixes: Suffixes
+    left: Index,
+    right: Index,
+    left_suffix: str | None,
+    right_suffix: str | None,
+    suffix_check_dups: bool = True,
 ) -> tuple[Index, Index]:
     """
     Suffixes type validation.
@@ -2705,17 +2751,11 @@ def _items_overlap_with_suffix(
     If corresponding suffix is empty, the entry is simply converted to string.
 
     """
-    if not is_list_like(suffixes, allow_sets=False) or isinstance(suffixes, dict):
-        raise TypeError(
-            f"Passing 'suffixes' as a {type(suffixes)}, is not supported. "
-            "Provide 'suffixes' as a tuple instead."
-        )
-
     to_rename = left.intersection(right)
     if len(to_rename) == 0:
         return left, right
 
-    lsuffix, rsuffix = suffixes
+    lsuffix, rsuffix = left_suffix, right_suffix
 
     if not lsuffix and not rsuffix:
         raise ValueError(f"columns overlap but no suffix specified: {to_rename}")
@@ -2746,17 +2786,20 @@ def _items_overlap_with_suffix(
     llabels = left._transform_index(lrenamer)
     rlabels = right._transform_index(rrenamer)
 
-    dups = []
-    if not llabels.is_unique:
-        # Only warn when duplicates are caused because of suffixes, already duplicated
-        # columns in origin should not warn
-        dups = llabels[(llabels.duplicated()) & (~left.duplicated())].tolist()
-    if not rlabels.is_unique:
-        dups.extend(rlabels[(rlabels.duplicated()) & (~right.duplicated())].tolist())
-    if dups:
-        raise MergeError(
-            f"Passing 'suffixes' which cause duplicate columns {set(dups)} is "
-            f"not allowed.",
-        )
+    if suffix_check_dups:
+        dups = []
+        if not llabels.is_unique:
+            # Only warn when duplicates are caused because of suffixes, already
+            # duplicated columns in origin should not warn
+            dups = llabels[(llabels.duplicated()) & (~left.duplicated())].tolist()
+        if not rlabels.is_unique:
+            dups.extend(
+                rlabels[(rlabels.duplicated()) & (~right.duplicated())].tolist()
+            )
+        if dups:
+            raise MergeError(
+                f"Passing 'suffixes' which cause duplicate columns {set(dups)} "
+                f"is not allowed.",
+            )
 
     return llabels, rlabels
